### PR TITLE
Added partially returned results - all connectors in async

### PR DESCRIPTION
**Pull request by mingazhev**
This new implementation allows to return results to client when at least one of connectors finished it's availability search.
Results for each connector are stored independently and the common result is assembled from them at runtime.

### CODE REVIEW

File: Api/Services/Accommodations/Availability/AvailabilityStorage.cs [hunk 1]
```diff
@@ -1,51 +1,134 @@
...
-                : result;
+            var providerSearchStates = await GetAll<AvailabilitySearchState>(searchId);
+            var successfulStatuses = providerSearchStates
+                .Select(s => s.Result.TaskState)
+                .Where(s => s != AvailabilitySearchTaskState.Unknown && s != AvailabilitySearchTaskState.Failed)
+                .ToHashSet();
+
+            if (successfulStatuses.Count == 0)
+            {
+                var error = string.Join(";", providerSearchStates.Select(p => p.Result.Error).ToArray());
```
Comment (oleg1std) [inline Api/Services/Accommodations/Availability/AvailabilityStorage.cs:82]: Space character is missing: "; "

File: Api/Services/Accommodations/Availability/AvailabilityStorage.cs [hunk 2]
```diff
@@ -1,51 +1,134 @@
...
+                return AvailabilitySearchState.Failed(searchId, error);
+            }
+                
+            var resultCount = providerSearchStates.Sum(s => s.Result.ResultCount);
+            if (successfulStatuses.Count == 1)
+            {
+                return AvailabilitySearchState.FromState(searchId, successfulStatuses.Single(), resultCount);
+            }
+            if(successfulStatuses.Contains(AvailabilitySearchTaskState.Completed))
```
Comment (oleg1std) [inline Api/Services/Accommodations/Availability/AvailabilityStorage.cs:91]: Space character is missing

File: Api/Services/Accommodations/Availability/MultiProviderAvailabilitySearchScheduler.cs [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+using Microsoft.AspNetCore.Mvc;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Logging;
+using AvailabilityRequest = HappyTravel.Edo.Api.Models.Availabilities.AvailabilityRequest;
+
+namespace HappyTravel.Edo.Api.Services.Accommodations.Availability
+{
+    public class MultiProviderAvailabilitySearchScheduler
+    {
+        public MultiProviderAvailabilitySearchScheduler(IServiceScopeFactory serviceScopeFactory,
```
Comment (oleg1std) [inline Api/Services/Accommodations/Availability/MultiProviderAvailabilitySearchScheduler.cs:22]: Is it possible to use here a shortened name, e.g. AvailabilitySearchScheduler

File: Api/Services/Accommodations/Availability/AvailabilityStorage.cs [hunk 4]
```diff
@@ -1,51 +1,134 @@
...
+            if (successfulStatuses.Count == 1)
+            {
+                return AvailabilitySearchState.FromState(searchId, successfulStatuses.Single(), totalResultsCount);
+            }
+            if(successfulStatuses.Contains(AvailabilitySearchTaskState.Completed))
+            {
+                return AvailabilitySearchState.PartiallyCompleted(searchId, totalResultsCount);
+            }   
+            
+            throw new ArgumentException($"Invalid tasks state: {string.Join(";", successfulStatuses)}");
```
Comment (oleg1std) [inline Api/Services/Accommodations/Availability/AvailabilityStorage.cs:96]: Space character

File: Api/Services/Accommodations/Availability/MultiProviderAvailabilitySearchScheduler.cs [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+
+                return new EdoContracts.Accommodations.AvailabilityRequest(request.Nationality, request.Residency, request.CheckInDate,
+                    request.CheckOutDate,
+                    request.Filters, roomDetails,
+                    new EdoContracts.GeoData.Location(location.Name, location.Locality, location.Country, location.Coordinates, location.Distance,
+                        location.Source, location.Type),
+                    request.PropertyType, request.Ratings);
+            }
+        }
+
```
Comment (oleg1std) [inline Api/Services/Accommodations/Availability/MultiProviderAvailabilitySearchScheduler.cs:87]: Extra line needed here

File: Api/Services/Accommodations/Availability/MultiProviderAvailabilitySearchScheduler.cs [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+            _logger.LogInformation($"Availability search tasks for search id '{searchId}' started");
+            
+            IReadOnlyCollection<(DataProviders Key, IDataProvider Provider)> GetProviders(in Location location)
+            {
+                var providers = location.DataProviders != null && location.DataProviders.Any()
+                    ? _dataProviderFactory.Get(location.DataProviders)
+                    : _dataProviderFactory.GetAll();
+                return providers;
+            }
+
```
Comment (oleg1std) [inline Api/Services/Accommodations/Availability/MultiProviderAvailabilitySearchScheduler.cs:72]: need one more line

File: Api/Services/Accommodations/Availability/AvailabilityStorage.cs (hunk 1 above)
Comment (mingazhev) [reply Api/Services/Accommodations/Availability/AvailabilityStorage.cs:82]: Fixed

File: Api/Services/Accommodations/Availability/AvailabilitySearchState.cs [hunk 7]
```diff
@@ -1,22 +1,67 @@
...
+using HappyTravel.Edo.Api.Models.Availabilities;
 using Newtonsoft.Json;
 
 namespace HappyTravel.Edo.Api.Services.Accommodations.Availability
 {
     public readonly struct AvailabilitySearchState
     {
         [JsonConstructor]
-        public AvailabilitySearchState(Guid id, AvailabilitySearchTaskState taskState, int resultCount = 0, string error = null)
+        private AvailabilitySearchState(Guid id, AvailabilitySearchTaskState taskState, int resultCount = 0, string error = null)
```
[the remark this answers is not in the document]
Comment (mingazhev) [reply Api/Services/Accommodations/Availability/AvailabilitySearchState.cs:10]: Do you mean serialization? Yes, looks like it works fine

File: Api/Services/Accommodations/Availability/AvailabilityStorage.cs [hunk 8]
```diff
@@ -1,51 +1,148 @@
...
+                    _memoryFlow.Set(key, providerSearchResults, CacheExpirationTime);
+            }
+
+            return CombineAvailabilities(providerSearchResults, page, pageSize);
+
+
+            static CombinedAvailabilityDetails CombineAvailabilities(List<(DataProviders ProviderKey, AvailabilityDetails Availability)> availabilities, int page, int pageSize)
+            {
+                if (availabilities == null || !availabilities.Any())
+                    return new CombinedAvailabilityDetails(default, default, default, default, default);
```
[the remark this answers is not in the document]
Comment (mingazhev) [reply Api/Services/Accommodations/Availability/AvailabilityStorage.cs:52]: Replaced with static property. For now we do not store request properties there.

File: Api/Services/Accommodations/Availability/AvailabilityStorage.cs [hunk 9]
```diff
@@ -1,51 +1,165 @@
...
+
+
+            static string GetErrors((DataProviders DataProvider, AvailabilitySearchState Result)[] states)
+            {
+                var errors = states
+                    .Select(p => p.Result.Error)
+                    .Where(e => !string.IsNullOrWhiteSpace(e))
+                    .ToArray();
+                
+                return string.Join(";", errors);
```
[the remark this answers is not in the document]
Comment (mingazhev) [reply Api/Services/Accommodations/Availability/AvailabilityStorage.cs:116]: Fixed

File: Api/Services/Accommodations/Availability/AvailabilityStorage.cs [hunk 10]
```diff
@@ -1,51 +1,148 @@
...
+                                    r.AccommodationDetails,
+                                    r.RoomContractSets);
+
+                                return ProviderData.Create(providerKey, result);
+                            })
+                            .ToList();
+
+                        return availabilityResults;
+                    })
+                    .Skip(pageSize * page)
```
[the remark this answers is not in the document]
Comment (mingazhev) [reply Api/Services/Accommodations/Availability/AvailabilityStorage.cs:74]: This is hard to realize for me now. To make it reproduceable we should add something like timestamp there (for each provider result).

File: Api/Infrastructure/Logging/LoggerExtensions.cs [hunk 11]
```diff
@@ -128,6 +129,22 @@ static LoggerExtensions()
             LocationNormalized = LoggerMessage.Define<string>(LogLevel.Information,
                 new EventId((int) LoggerEvents.LocationNormalized, LoggerEvents.LocationNormalized.ToString()),
                 $"INFORMATION | {nameof(LocationNormalizer)}: {{message}}");
+            
+            AvailabilitySearchStartedEventOccured = LoggerMessage.Define<string>(LogLevel.Information,
```
[the remark this answers is not in the document]
Comment (mingazhev) [reply Api/Infrastructure/Logging/LoggerExtensions.cs:133]: Changed to TRACE

File: Api/Services/Accommodations/Availability/AvailabilityStorage.cs (hunk 10 above)
[the remark this answers is not in the document]
Comment (kirillta) [reply Api/Services/Accommodations/Availability/AvailabilityStorage.cs:74]: Ok then, let's put the issue aside for now

File: Api/Controllers/AccommodationsController.cs [hunk 12]
```diff
@@ -109,30 +87,32 @@ public async Task<IActionResult> StartAvailabilitySearch([FromBody] Availability
...
+        /// <param name="page">Page number</param>
+        /// <param name="pageSize">Page size</param>
+        /// <returns>Availability results</returns>
+        [HttpGet("availabilities/accommodations/searches/{searchId}")]
         [ProducesResponseType(typeof(CombinedAvailabilityDetails), (int) HttpStatusCode.OK)]
         [ProducesResponseType(typeof(ProblemDetails), (int) HttpStatusCode.BadRequest)]
         [MinCounterpartyState(CounterpartyStates.ReadOnly)]
         [InAgencyPermissions(InAgencyPermissions.AccommodationAvailabilitySearch)]
-        public async Task<IActionResult> GetAvailabilitySearchResult([FromRoute] Guid searchId)
+        public async Task<IActionResult> GetAvailabilitySearchResult([FromRoute] Guid searchId, [FromQuery] int? page, [FromQuery] int? pageSize)
```
Comment (kirillta) [inline Api/Controllers/AccommodationsController.cs:112]: Add defaults to the page and the pageSize, and why it's not in a take/skip style?

File: Api/Services/Accommodations/Availability/IAvailabilityStorage.cs [hunk 13]
```diff
@@ -0,0 +1,19 @@
...
+using System.Threading.Tasks;
+using HappyTravel.Edo.Api.Models.Accommodations;
+using HappyTravel.Edo.Common.Enums;
+using HappyTravel.EdoContracts.Accommodations;
+
+namespace HappyTravel.Edo.Api.Services.Accommodations.Availability
+{
+    public interface IAvailabilityStorage
+    {
+        Task SaveResult(Guid searchId, DataProviders dataProvider, AvailabilityDetails details);
```
Comment (kirillta) [inline Api/Services/Accommodations/Availability/IAvailabilityStorage.cs:11]: Why not `SetState` like in other classes?

Comment (mingazhev) [reply Api/Services/Accommodations/Availability/IAvailabilityStorage.cs:11]: SetState sounds better, agree

File: Api/Controllers/AccommodationsController.cs (hunk 12 above)
Comment (mingazhev) [reply Api/Controllers/AccommodationsController.cs]: changed to top and skip

**Pull request merged**